extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeThrottlingDecrypter.java

```diff
@@ -0,0 +1,95 @@
+ * which is not throttled.
+ * </p>
+ *
+ * <p>
+ * https://r5---sn-4g5ednsz.googlevideo.com/videoplayback?n=VVF2xyZLVRZZxHXZ&other=other
+ * </p>
+ * becomes
+ * <p>
+ * https://r5---sn-4g5ednsz.googlevideo.com/videoplayback?n=iHywZkMipkszqA&other=other
+ * </p>
+ */
+public class YoutubeThrottlingDecrypter {
+
+    private static final String N_PARAM_REGEX = "[&?]n=([^&]+)";
+    private static final Map<String, String> nParams = new HashMap<>();
+
+    private final String functionName;
+    private final String function;
+
+    /**
+     * <p>
+     * Use this if you care about the off chance that YouTube tracks with which videoId the cipher is requested.
+     * </p>
+     * Otherwise use the no-arg constructor which uses a constant value.
+     */
+    public YoutubeThrottlingDecrypter(String videoId) throws ParsingException {
+        final String playerJsCode = YoutubeJavaScriptExtractor.extractJavaScriptCode(videoId);
+
+        functionName = parseDecodeFunctionName(playerJsCode);
+        function = parseDecodeFunction(playerJsCode, functionName);
+    }
+
+    public YoutubeThrottlingDecrypter() throws ParsingException {
+        final String playerJsCode = YoutubeJavaScriptExtractor.extractJavaScriptCode();
+
+        functionName = parseDecodeFunctionName(playerJsCode);
+        function = parseDecodeFunction(playerJsCode, functionName);
+    }
+
+    private String parseDecodeFunctionName(String playerJsCode) throws Parser.RegexException {
+        Pattern pattern = Pattern.compile("b=a\\.get\\(\"n\"\\)\\)&&\\(b=(\\w+)\\(b\\),a\\.set\\(\"n\",b\\)");
+        return Parser.matchGroup1(pattern, playerJsCode);
+    }
+
+    private String parseDecodeFunction(String playerJsCode, String functionName) throws Parser.RegexException {
+        Pattern functionPattern = Pattern.compile(functionName + "=function(.*?;)\n", Pattern.DOTALL);
+        return  "function " + functionName + Parser.matchGroup1(functionPattern, playerJsCode);
+    }
+
+    public String apply(String url) throws Parser.RegexException {
+        if (containsNParam(url)) {
+            String oldNParam = parseNParam(url);
+            String newNParam = decryptNParam(oldNParam);
+            return replaceNParam(url, oldNParam, newNParam);
+        } else {
+            return url;
+        }
+    }
+
+    private boolean containsNParam(String url) {
+        return Parser.isMatch(N_PARAM_REGEX, url);
+    }
+
+    private String parseNParam(String url) throws Parser.RegexException {
+        Pattern nValuePattern = Pattern.compile(N_PARAM_REGEX);
+        return Parser.matchGroup1(nValuePattern, url);
+    }
+
+    private String decryptNParam(String nParam) {
+        if (nParams.containsKey(nParam)) {
+            return nParams.get(nParam);
+        }
+        final String decryptedNParam = JavaScript.run(function, functionName, nParam);
+        nParams.put(nParam, decryptedNParam);
```

Comment:
Looks like ``nParams`` is used for caching.
Shouldn't that be cleaned up after a while (my fear of memory leaks kicks in 😆)?
Or at least be shomehow reachable from NewPipe when clicking ``Wipe cached metadata``?

Reply:
We store one nParam per video. that not much when using the extractor in the app, because the cache is reset when closing the application. I wouldn't worry too much about that.
When using the extractor in a different environment like Piped, the cache might grow fast. I added two methods `getCacheSize()` and `clearCache()`. Maybe @FireMasterK can give us some stats about the size of the cache later.

Reply:
Alternatively, if you use a library like https://github.com/ben-manes/caffeine, you can evict old entries based on size or time too.

Reply:
I think `getCacheSize()` and `clearCache()` are enough for now. The jar of caffeine seems to be nearly 1mb, which i think would bloat the library too much, for the actual functionality that we need from it.